ui/tests/acceptance/mfa-test.js

```diff
@@ -0,0 +1,135 @@
+
+  const login = async (user) => {
+    // MfaHandler(server);
+    await visit('/vault/auth');
+    await fillIn('[data-test-select="auth-method"]', 'userpass');
+    await fillIn('[data-test-username]', user);
+    await fillIn('[data-test-password]', 'test');
+    await click('[data-test-auth-submit]');
+  };
+  const didLogin = (assert) => {
+    assert.equal(currentRouteName(), 'vault.cluster.secrets.backends', 'Route transitions after login');
+  };
+  const validate = async (multi) => {
+    await fillIn('[data-test-mfa-passcode="0"]', 'test');
+    if (multi) {
+      await fillIn('[data-test-mfa-passcode="1"]', 'test');
+    }
+    await click('[data-test-mfa-validate]');
+  };
+
+  test('it should handle single mfa constraint with passcode method', async function (assert) {
+    await login('mfa-a');
+    assert
+      .dom('[data-test-mfa-description]')
+      .includesText(
+        'Enter your authentication code to log in.',
+        'Mfa form displays with correct description'
+      );
+    assert.dom('[data-test-mfa-select]').doesNotExist('Select is hidden for single method');
+    assert.dom('[data-test-mfa-passcode]').exists({ count: 1 }, 'Single passcode input renders');
+    await validate();
+    didLogin(assert);
+  });
+
+  test('it should handle single mfa constraint with push method', async function (assert) {
+    await login('mfa-b');
+    didLogin(assert);
+  });
+
+  test('it should handle single mfa constraint with 2 passcode methods', async function (assert) {
+    await login('mfa-c');
+    assert
+      .dom('[data-test-mfa-description]')
+      .includesText('Select the MFA method you wish to use.', 'Mfa form displays with correct description');
+    assert
+      .dom('[data-test-mfa-select]')
+      .exists({ count: 1 }, 'Select renders for single constraint with multiple methods');
+    assert.dom('[data-test-mfa-passcode]').doesNotExist('Passcode input hidden until selection is made');
+    await this.select();
+    await validate();
+    didLogin(assert);
+  });
+
+  test('it should handle single mfa constraint with 2 push methods', async function (assert) {
+    await login('mfa-d');
+    await this.select();
+    await click('[data-test-mfa-validate]');
+    didLogin(assert);
+  });
+
+  test('it should handle single mfa constraint with 1 passcode and 1 push method', async function (assert) {
+    await login('mfa-e');
+    await this.select(0, 2);
+    assert.dom('[data-test-mfa-passcode]').exists('Passcode input renders');
+    await this.select();
+    assert.dom('[data-test-mfa-passcode]').doesNotExist('Passcode input is hidden for push method');
+    await click('[data-test-mfa-validate]');
+    didLogin(assert);
+  });
+
+  test('it should handle multiple mfa constraints with 1 passcode method each', async function (assert) {
+    await login('mfa-f');
+    assert
+      .dom('[data-test-mfa-description]')
+      .includesText(
+        '2 methods are required for successful authentication.',
```

Comment:
not sure where this is in the code, but if possible it seems like this should be a written two instead of 2. Any numbers under 10 are generally written out.

Reply:
I think a helper for this would be useful! Angel is right, usually numbers under 10 are wordy